File: apps/api/app/schemas/ai.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class ExplainIn(BaseModel):
    """Ask the AI to explain a question's answer."""

    question_id: int
    chosen_option_id: int | None = None
    text_answer: str | None = None
    locale: str = Field(default="en", max_length=8)


class ExplainOut(BaseModel):
    explanation: str
    # None when the question is self-assessed (writing/speaking).
    is_correct: bool | None = None
    correct_answer: str | None = None
    # True when the text came from Gemini, False when from the dev fallback.
    ai_generated: bool = False


class TutorChatIn(BaseModel):
    message: str = Field(min_length=1, max_length=2000)
    locale: str = Field(default="en", max_length=8)


class TutorMessageOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    role: str
    content: str
    created_at: datetime


class TutorChatOut(BaseModel):
    reply: str
    ai_generated: bool = False


class StudyPlanItem(BaseModel):
    section_id: int
    section_title: str
    level_code: str
    skill_name: str
    answered: int
    accuracy: int
    reason: str


class StudyPlanOut(BaseModel):
    summary: str
    recommendations: list[StudyPlanItem]
    ai_generated: bool = False

File: apps/web/src/lib/ai.ts
const API_BASE =
  process.env.NEXT_PUBLIC_API_BASE_URL ?? "http://localhost:8000";
const API_V1 = `${API_BASE}/api/v1`;

async function call<T>(
  path: string,
  options: RequestInit & { token?: string } = {}
): Promise<T> {
  const { token, headers, ...rest } = options;
  const res = await fetch(`${API_V1}${path}`, {
    ...rest,
    headers: {
      "Content-Type": "application/json",
      ...(token ? { Authorization: `Bearer ${token}` } : {}),
      ...headers
    }
  });
  if (!res.ok) {
    throw new Error(`AI request failed (${res.status})`);
  }
  return (await res.json()) as T;
}

export interface ExplainResult {
  explanation: string;
  is_correct: boolean | null;
  correct_answer: string | null;
  ai_generated: boolean;
}

export interface TutorMessage {
  role: "user" | "assistant";
  content: string;
  created_at?: string;
}

export interface TutorReply {
  reply: string;
  ai_generated: boolean;
}

export interface StudyPlanItem {
  section_id: number;
  section_title: string;
  level_code: string;
  skill_name: string;
  answered: number;
  accuracy: number;
  reason: string;
}

export interface StudyPlan {
  summary: string;
  recommendations: StudyPlanItem[];
  ai_generated: boolean;
}

export const aiApi = {
  status: () => call<{ ai_enabled: boolean }>("/ai/status"),

  explain: (
    params: {
      question_id: number;
      chosen_option_id?: number | null;
      text_answer?: string | null;
      locale?: string;
    },
    token?: string
  ) =>
    call<ExplainResult>("/ai/explain", {
      method: "POST",
      body: JSON.stringify(params),
      token
    }),

  chat: (message: string, locale: string, token: string) =>
    call<TutorReply>("/ai/tutor/chat", {
      method: "POST",
      body: JSON.stringify({ message, locale }),
      token
    }),

  history: (token: string) =>
    call<TutorMessage[]>("/ai/tutor/history", { token }),

  studyPlan: (token: string) => call<StudyPlan>("/ai/study-plan", { token })
};
